compiler/luci/service/src/Nodes/CircleFullyConnected.cpp

```diff
@@ -36,4 +40,70 @@ luci::CircleNode *CloneNodeLet<CN::DEF>::visit(const luci::CircleFullyConnected
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleFullyConnected *node)
+{
+  auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->input()));
+  auto weights_shape = circle_shape(loco::must_cast<CircleNode *>(node->weights()));
+
+  loco::TensorShape out_shape;
+
+  // NOTE Some recipes in some repositories are using rank 4 input for FullyConnected.
+  //      Until they are all fixed, disable following assert.
+  // TODO Enable following assert after related fixes are applied
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L194
+  // LUCI_ASSERT(input_shape.rank() == 2 || input_shape.rank() == 3,
+  //             "Input rank of FullyConnected should be 2 or 3");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L225
+  LUCI_ASSERT(weights_shape.rank() == 2, "Weights of FullyConnected should be 2");
+
```

Comment:
IMHO, we need to handle this case. 

e.g. `input = {1, 2, 4, 3}` `weights ={5, 4}` - when `input[-1]` and `weights[-1]` are both known but it holds different values, (I guess) we need to throw an error in this case. 

```suggestion

   const auto o_shape_last = out_shape.dim(out_shape.rank()-1);
   const auto w_shape_last = weights_shape.dim(1);
   if (o_shape_last.known() && w_shape_last.known() && not (o_shape_last != w_shape_last))
   {
   	// throw error 
   }
```

Reply:
I agree.
I've considered this case, and I believe it should be separated from the dynamic shape handling.
Currently, I'm looking into how TensorFlow Lite determines the weights in a fully connected layer. Is there a function that checks whether the weights are set correctly, as in this case?
If not, we might need to throw an error as you suggested.

Reply:
>  it should be separated from the dynamic shape handling. 

You're right. 

:thinking: But IMO, this is not related to the **unknown** shape, still it is related to shape inference. While shape inferencing, 'input = {1, 4, 3} weights = {5, 4}' comes as the input, shape inference should explicitly throw an error, instead of inferring the wrong output shape.

Reply:
I reconsider your opinion, I also agree with your point :) 

Such invalid graph means that a graph was created incorrectly from the beginning.
Checking an Invalid graph is not part of the shape inference.

Reply:
> Currently, I'm looking into how TensorFlow Lite determines the weights in a fully connected layer. Is there a function that checks whether the weights are set correctly, as in this case? 

I checked XLA(backend compiler of Tensorflow) - https://github.com/tensorflow/tensorflow/blob/ad6d8cc177d0c868982e39e0823d0efbfb95f04c/third_party/xla/xla/service/shape_inference.cc#L844, 

But hard to distinguish it checks above cases or not :cry:

Reply:
> I reconsider your opinion, I also agree with your point :)
> 
> Such invalid graph means that a graph was created incorrectly from the beginning. Checking an Invalid graph is not part of the shape inference.
> 
> +) I think you can decide whether to throw an error or not. It's up to your decision :)

Even if it doesn't apply to inference, I think throwing an error is a good approach.